### ISSUE TITLE
stbi_write_jpg / stbi_load incorrect channel count in monochrome images 

### ISSUE THREAD

**Issue by RafigRzayev7** [nothings/stb#889]
I have constructed data for monochrome image inside a program. 

Then I saved it as a single channel jpg:
 `stbi_write_jpg( "filename", W, H, 1, data, 100) // saved as 1 channel`

Afterwards, I loaded the saved image with expectation that it will be single channel: 
`stbi_load("filename", &w, &h, &channels_, 1); // tried both 0 and 1 in last argument`

**Value of channels_ should be set to 1, but it is always 3.** 

It seems to me as a bug. 

I found this behavior in my application and it causes serious problems. If it is normal behavior could you please tell me why and share if there is a simple way to infer that `channels_ == 1` ?

I prepared small C++ test program to demonstrate this behavior: 
```
#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
#include <iostream>

int main() {
// Generate monochrome image 
const int WIDTH{5}, HEIGHT{4}, CHANNELS{1};
unsigned char* data = new unsigned char[WIDTH*HEIGHT*CHANNELS] {
  25, 50, 100, 150, 200,
  35, 60, 110, 160, 210,
  45, 70, 120, 170, 220,
  55, 80, 130, 180, 230
};
// Save image as monochrome
int save_status =
    stbi_write_jpg("new_save.jpg", WIDTH, HEIGHT, CHANNELS, data, 100);
if (save_status == 0) {
  std::cout << "ISSUE DURING SAVE\n";
}
// Show generated image dimensions & colors
std::cout << "GENERATED MONOCHROME IMAGE:\n";
std::cout << "Width: " << WIDTH << std::endl;
std::cout << "Height: " << HEIGHT << std::endl;
std::cout << "Channels: " << CHANNELS << std::endl;

// Load saved image
int loaded_width{}, loaded_height{}, loaded_channels{};
unsigned char *loaded_data = stbi_load("new_save.jpg", &loaded_width,
                                       &loaded_height, &loaded_channels, 1);
if (loaded_data == NULL) {
  std::cout << "ISSUE DURING LOAD\n";
}
// Show image dimensions & colors
std::cout << "\nLOADED MONOCHROME IMAGE:\n";
std::cout << "Width: " << loaded_width << std::endl;
std::cout << "Height: " << loaded_height << std::endl;
std::cout << "Channels: " << loaded_channels << std::endl;

return 0;
}
```

**Comment by nothings** [comment]
I didn't write or integrate the jpeg writer, so somebody needs to read the code and figure out what it's doing. Maybe it just doesn't support monochrome, I don't know.

**Comment by RafigRzayev7** [comment]
I tried stbi_load() to load a gray image on my PC, composition value was correct (1).

**Problem is in stbi_write_jpg()**. I looked into code, though I am not very familiar with the code and JPEG format, I guess there is no specific implementation for monochrome. If I don't mistake YUV data is being written, and we need to write just Y data. 

Can you help @DanielGibson ?

**Comment by DanielGibson** [comment]
Does the jpeg file look correct in an image viewer?

stbi_write_jpg() always writes RGB data (monochrome data is converted)

**Comment by nothings** [comment]
Ok, that's inconsistent with how all of the other stb_image_write implementations work--they all write out data as it's provided. JPEG does support monochrome; does olick's jpeg writer not support it?

**Comment by DanielGibson** [comment]
As far as I understand, it doesn't support it.
It just gets r, g, b values from the same input byte instead of different ones (as it does for rgb(a) input) and otherwise does the same everywhere else, unless I miss something:
https://github.com/nothings/stb/blob/master/stb_image_write.h#L1468

**Comment by rygorous** [comment]
Yeah, the JPEG writer always writes RGB JPEGs no matter what. This would take changes to several places in the JPEG writer. I'm classifying this as an enhancement request, it's not a trivial fix.
